### PR TITLE
Support JDBC escape syntax

### PR DESCRIPTION
**Pull request by nicktorwald**
Add a driver SQL pre-processing before sending it to the server. The
driver supports sub-set of scalar functions defined by the spec
(appendix C), outer joins, escape clause for SQL LIKE operator, and
limit/offset clause. The processed result can be received using
Connection.nativeSQL() method.

Closes: #79
Closes: #76
Closes: #81
Closes: #83
Closes: #84
Affects: #108

### CODE REVIEW

File: src/main/java/org/tarantool/jdbc/EscapedFunctions.java [hunk 1]
```diff
@@ -0,0 +1,412 @@
...
+ *     </tr>
+ *     <tr>
+ *         <td>REPLACE(string1, string2, string3)</td>
+ *         <td>REPLACE(string1, string2, string3)</td>
+ *         <td></td>
+ *     </tr>
+ *     <tr>
+ *         <td>RIGHT(string, count)</td>
+ *         <td>SUBSTR(string, LENGTH(string) - count + 1)</td>
+ *         <td>Can produce too long native expression because string is used twice</td>
```
Comment (Totktonada) [inline src/main/java/org/tarantool/jdbc/EscapedFunctions.java]: `SUBSTR('abcd', -2)` gives `'cd'`: we can eliminate double using of a string.

File: src/main/java/org/tarantool/jdbc/EscapedFunctions.java [hunk 2]
```diff
@@ -0,0 +1,412 @@
...
+ *     <caption>Supported system scalar functions</caption>
+ *     <tr>
+ *         <th>JDBC escape</th>
+ *         <th>Native</th>
+ *         <th>Comment</th>
+ *     </tr>
+ *     <tr>
+ *         <td>DATABASE()</td>
+ *         <td>'DEFAULT'</td>
+ *         <td>Tarantool does not support databases. Driver always replaces it to 'DEFAULT'.</td>
```
Comment (Totktonada) [inline src/main/java/org/tarantool/jdbc/EscapedFunctions.java]: Nit: As proposed in PR #225, I would return `'universe'` here.

File: src/main/java/org/tarantool/jdbc/EscapedFunctions.java [hunk 3]
```diff
@@ -0,0 +1,412 @@
...
+ *     </tr>
+ *     <tr>
+ *         <td>IFNULL(expression1, expression2)</td>
+ *         <td>IFNULL(expression1, expression2)</td>
+ *         <td></td>
+ *     </tr>
+ *     <tr>
+ *         <td>USER()</td>
+ *         <td>'GUEST'</td>
```
Comment (Totktonada) [inline src/main/java/org/tarantool/jdbc/EscapedFunctions.java]: Nit: `'GUEST'` is this example should be in lower case, I guess.

File: src/main/java/org/tarantool/jdbc/EscapeSyntaxParser.java [hunk 4]
```diff
@@ -0,0 +1,319 @@
...
+                        endOfComment = sql.length() - 1;
+                    }
+                }
+                if (i == endOfComment) {
+                    activeBuffer.append(currentChar);
+                    i++;
+                } else {
+                    if (activeBuffer == nativeSql && preserveComments) {
+                        nativeSql.append(sql, i, endOfComment + 1);
+                    }
```
[the remark this answers is not in the document]
Comment (nicktorwald) [reply src/main/java/org/tarantool/jdbc/EscapeSyntaxParser.java]: removed

File: src/main/java/org/tarantool/jdbc/EscapeSyntaxParser.java [hunk 5]
```diff
@@ -0,0 +1,319 @@
...
+                }
+                i++;
+                break;
+
+            case ')':
+                if (--braceNestLevel == 0) {
+                    // reach a function closing brace
+                    // parse the last possible function parameter
+                    String param = functionString.substring(parameterStartPosition, i).trim();
+                    if (!functionParameters.isEmpty() || !param.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (nicktorwald) [reply src/main/java/org/tarantool/jdbc/EscapeSyntaxParser.java]: reorganized conditions to check empty arguments `{fn substring('abcd', 1, )}` (the third argument is empty and should cause an exception)

File: src/main/java/org/tarantool/jdbc/EscapeSyntaxParser.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (Totktonada) [reply src/main/java/org/tarantool/jdbc/EscapeSyntaxParser.java]:
```diff
@@ -101,9 +104,6 @@ public class EscapeSyntaxParser {
                     activeBuffer.append(currentChar);
                     i++;
                 } else {
-                    if (activeBuffer == nativeSql && preserveComments) {
-                        nativeSql.append(sql, i, endOfComment + 1);
-                    }
                     i = endOfComment + 1;
                 }
                 break;
```

So it seems that the problem with `token1/*comment*/token2` still here. We should add a whitespace or a newline if we'll strip comments.

Aside of that we discussed this a bit with @Korablev77 (Tarantool/SQL guy) and we both doubt that stripping of comments is the good idea if it is not required by the standard and is not requested by a user. The point here is to provide good debugging facilities: say, report correct line number and full text of a request in case of an error.

I propose to don't strip anything by default and maybe add an option in a future (if we'll see a demand).

BTW, why comments in escape syntax are stripped unconditionally? I would say that we should not strip them too as well (but I don't know if there any problems with it).

[the remark this answers is not in the document]
Comment (Totktonada) [reply src/main/java/org/tarantool/jdbc/EscapeSyntaxParser.java]: Hm, AFAIS, now escape syntax comments are preserved unconditionally. So only the question remains is about comments outside of escape syntax.

File: src/main/java/org/tarantool/jdbc/EscapeSyntaxParser.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (Totktonada) [reply src/main/java/org/tarantool/jdbc/EscapeSyntaxParser.java]: Looks correct now.

File: src/main/java/org/tarantool/jdbc/EscapeSyntaxParser.java [hunk 6]
```diff
@@ -0,0 +1,319 @@
...
+                break;
+            }
+        }
+
+        if (functionName == null || braceNestLevel != 0) {
+            throw new SQLSyntaxErrorException(
+                "Malformed function expression '" + functionString + "'", SQLStates.SYNTAX_ERROR.getSqlState()
+            );
+        }
+        return new FunctionExpression(functionName, functionParameters);
```
[the remark this answers is not in the document]
Comment (Totktonada) [reply src/main/java/org/tarantool/jdbc/EscapeSyntaxParser.java]: A tail is checked now — okay.

File: src/main/java/org/tarantool/jdbc/EscapeSyntaxParser.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (Totktonada) [reply src/main/java/org/tarantool/jdbc/EscapeSyntaxParser.java]: > Hm, AFAIS, now escape syntax comments are preserved unconditionally. So only the question remains is about comments outside of escape syntax.

My bad, I misinterpreted the updated code. Yep, now comments are removed everywhere. My proposal is to preserve them everywhere.

[the remark this answers is not in the document]
Comment (nicktorwald) [reply src/main/java/org/tarantool/jdbc/EscapeSyntaxParser.java]: Agree, but I have one objection here. It touches comments within escape expressions. 

- I found out there are cases when we cannot definitely keep comments as-is. Let's say, an expression like `select {fn database(/* get db name */)}`; here the transformed query will be `select 'universe'`, thus, it makes a correct transformation impossible, IMO.

- Next, preserving comments is working now till we just omitting `{}` braces (oj, limit, etc.); but we cannot guarantee that escape syntax won't use a more complex transformation in the future that won't be compatible with comments as in the previous point. 

- JDBC does not allow or deny comments inside `{}` explicitly. But the syntax is more or less strict (i.e. the scalar function is described as `{fn <function-name> (argument list)}`) and I think here a more correct way is to inform users that we follow the escape syntax in a stricter way here and the comments within *escapes* are not supported. it might cause a syntax exception or silent filtering out. I've chosen the second option: write down attention in Javadoc (+JDCB doc) that SQL comments won't be present after the processing because they can impact on the correct transformation.

To sum it up, it's ok to preserve the comments outside the escapes (debugging and other purposes) but there are a few (lot?) corner cases if we try to handle comments in a block being
processed by the driver. I'd like to always replace them with whitespaces in this case.

An example: `select /*may return guest*/ {fn user(/*current user*)}` becomes `select /*may return guest*/ 'guest'`.

File: src/test/java/org/tarantool/jdbc/JdbcPreparedStatementIT.java [hunk 7]
```diff
@@ -719,6 +719,161 @@ public void testSetBadCharacterStream() throws Exception {
...
+            assertFalse(resultSet.next());
+        }
+    }
+
+    @Test
+    public void testStringFunctionEscapeSyntax() throws Exception {
+        testHelper.executeSql("INSERT INTO test(id, val) VALUES (1, 'one'), (2, 'TWO'), (3, 'three'), (4, ' four ')");
+
+        prep = conn.prepareStatement(
+            "SELECT {fn char(?)}, {fn right(val, 2)}, {fn concat(?, val)} FROM test WHERE id = 3"
```
[the remark this answers is not in the document]
Comment (nicktorwald) [reply src/test/java/org/tarantool/jdbc/JdbcPreparedStatementIT.java]: well, we can't use `existing LENGTH() function` directly because of incompatible with the spec behavior;

Appendix C says:

> LENGTH(string) - Number of characters in a string, **excluding trailing blanks**

This is the reason we transform it into `LENGTH(TRIM(TRAILING FROM string))`. But `TRAILING FROM` is not supported in 2.1 too.

I moved a few v2.1-unsupported functions into separate tests.

https://github.com/tarantool/tarantool-java/blob/d1f72f28ea72b18d729774c4f5849ce3033dd67f/src/test/java/org/tarantool/jdbc/JdbcPreparedStatementIT.java#L887-L909

https://github.com/tarantool/tarantool-java/blob/d1f72f28ea72b18d729774c4f5849ce3033dd67f/src/test/java/org/tarantool/jdbc/JdbcStatementIT.java#L707-L726

[the remark this answers is not in the document]
Comment (nicktorwald) [reply src/test/java/org/tarantool/jdbc/JdbcPreparedStatementIT.java]: One more change is related to `RAND` function:

> RAND(integer) - Random **floating-point** for seed integer

Here, I set eyes on `floating-point` that it usually means a value within [0, 1) (according to other jdbc vendors too). It is also not the same as Tarantool `RANDOM()` function which returns integer value between [-2^63-1, 2^64], AFAIK. I was thinking of how we can overcome and normalize this value to the required range, say using smt like `abs(random())/max_int` but failed due to the formula extra complexity. So, Instead of using `RANDOM()` the driver uses its own PRNG (`new Random().nextDouble()`) to substitute invocation of the function by double literal directly.

File: src/main/java/org/tarantool/jdbc/EscapeSyntaxParser.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (Totktonada) [reply src/main/java/org/tarantool/jdbc/EscapeSyntaxParser.java]: `/* <...> */` and `-- <...> EOL` comments cannot be misinterpreted as some other SQL clause and vice versa. The only case when we're unable to keep a comment as is is when we replace an entire expression that contains it on a driver side: this is the valid transformation. So I don't see how comments may impact correctness of a transformation.

JDBC is naturally based on SQL and, say, expression `{fn <function name> (<argument list>)` treats `<function name>` as identifier using SQL rules, tokenizes `(`, `)` and `<argument list>` as SQL requires, assumes that `<argument list>` is a list of valid comma separated SQL expressions. At least it means that `<argument list>` can contain a comment, but I think it is more natural to support comments for escape syntax in SQL way: on tokenizer level.

If we'll need to make more complex transformations, it would be better to work on extraction of SQL parser from tarantool (https://github.com/tarantool/tarantool/issues/3319) and transform an AST rather then a text on driver's side.

I have two possible reasons in the mind to keep comments:

* A user may have deep nested joins and may want to comment them.
* A kind of unification on tokenizer level: the same rules for SQL everywhere.

But we can look what other connectors doing. Maybe my way to think about those aspects is not popular.

File: src/test/java/org/tarantool/jdbc/JdbcPreparedStatementIT.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (Totktonada) [reply src/test/java/org/tarantool/jdbc/JdbcPreparedStatementIT.java]: Re `LTRIM()` and `RTRIM()`: it was changed in https://github.com/tarantool/tarantool/issues/3879 since 2.2.0-198-gf95a34da9 (and for all 2.3 versions). tarantool-2.1 has syntax inherited from SQLite: `LTRIM()` and `RTRIM()`. I would make the replacement based on tarantool version.

Okay for `RAND()`.
